crates/chain/src/local_chain.rs

```diff
@@ -264,43 +187,183 @@ impl LocalChain {
+        update: CheckPoint<D>,
+    ) -> Result<ChangeSet<D>, CannotConnectError> {
+        let (new_tip, changeset) = merge_chains(self.tip.clone(), update)?;
+        self.tip = new_tip;
+        Ok(changeset)
+    }
 
     /// Apply the given `changeset`.
-    pub fn apply_changeset(&mut self, changeset: &ChangeSet) -> Result<(), MissingGenesisError> {
+    pub fn apply_changeset(&mut self, changeset: &ChangeSet<D>) -> Result<(), MissingGenesisError> {
         let old_tip = self.tip.clone();
         let new_tip = apply_changeset_to_checkpoint(old_tip, changeset)?;
         self.tip = new_tip;
-        debug_assert!(self._check_changeset_is_applied(changeset));
         Ok(())
     }
 
-    /// Insert a [`BlockId`].
+    /// Derives an initial [`ChangeSet`], meaning that it can be applied to an empty chain to
+    /// recover the current chain.
+    pub fn initial_changeset(&self) -> ChangeSet<D> {
+        ChangeSet {
+            blocks: self
+                .tip
+                .iter()
+                .map(|cp| (cp.height(), Some(cp.data())))
+                .collect(),
+        }
+    }
+
+    /// Insert block into a [`LocalChain`].
     ///
     /// # Errors
     ///
     /// Replacing the block hash of an existing checkpoint will result in an error.
-    pub fn insert_block(&mut self, block_id: BlockId) -> Result<ChangeSet, AlterCheckPointError> {
-        if let Some(original_cp) = self.tip.get(block_id.height) {
+    pub fn insert_block(
+        &mut self,
+        height: u32,
+        data: D,
+    ) -> Result<ChangeSet<D>, AlterCheckPointError> {
+        if let Some(original_cp) = self.tip.get(height) {
             let original_hash = original_cp.hash();
-            if original_hash != block_id.hash {
+            if original_hash != data.to_blockhash() {
                 return Err(AlterCheckPointError {
-                    height: block_id.height,
+                    height,
                     original_hash,
-                    update_hash: Some(block_id.hash),
+                    update_hash: Some(data.to_blockhash()),
                 });
             }
             return Ok(ChangeSet::default());
         }
 
-        let mut changeset = ChangeSet::default();
-        changeset
-            .blocks
-            .insert(block_id.height, Some(block_id.hash));
+        let mut changeset = ChangeSet::<D>::default();
+        changeset.blocks.insert(height, Some(data));
         self.apply_changeset(&changeset)
             .map_err(|_| AlterCheckPointError {
                 height: 0,
                 original_hash: self.genesis_hash(),
-                update_hash: changeset.blocks.get(&0).cloned().flatten(),
+                update_hash: Some(
+                    changeset
+                        .blocks
+                        .get(&0)
+                        .cloned()
+                        .flatten()
+                        .unwrap()
+                        .to_blockhash(),
+                ),
```

Comment:
I spotted an `unwrap`, not sure if it was intentional.
```suggestion
                update_hash: changeset
                    .blocks
                    .get(&0)
                    .cloned()
                    .flatten()
                    .map(|d| d.to_blockhash()),
```